### DISCUSSION TITLE
How can I use Envoy's cluster specifier on all routes

### DISCUSSION THREAD

**Discussion by vladfr** [istio/istio#49117] [Q&A]
Envoy has a scriptable [cluster specifier plugin](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/cluster_specifier/lua#config-http-cluster-specifier-lua), where for each route you can have a Lua script that sets the destination cluster. 

How can I patch this with `EnvoyFilter` so that I add a cluster specifier script to all of my routes? I can't figure out a way to write the `EnvoyFilter` to allow for this. I believe this cannot be done with the current `patch` functionality.

For context, I would like to send requests to a different cluster based on a HTTP header.

**Comment by rootsongjc** [comment]
It seems to me that this is a matter of dynamic routing based on HTTP headers. Is the number of your destinations limited? If so, you can write a Lua script in EnvoyFilter by adding HTTP headers and then routing them in VirtualService based on the Header values.

**Comment by vladfr** [comment]
This is an interesting idea, to use another layer of VirtualService. Is is
needed because a HTTP filter can't alter the route destination. But, you
can just proxy by lua to achieve the same effect.

The number of destinations isn't limited. I basically want to allow devs to
deploy as many versions as they want, and route per-request, by just adding
headers (e.g.  to all requests made by the frontend)

I'm toying with different concepts, but I've identified the cluster
specifier as being the most straightforward thing to use. If only I could
patch it with EnvoyFilter!

**Comment by vladfr** [comment]
Coming back to this. I have a lua filter that reads headers, makes some changes, and makes a `httpCall` to another VirtualService fronting a `httpbin`. In the `httpbin` VS, I can route based on the headers. My example involved a `directResponse` on a header. However, this does not work. 

In my lua filter, I can get a reply from the `httpbin`, but the routing rules that I've applied in the VirtualService don't matter. If I call my `httpbin` service directly from a `sleep` pod, it works and I get my `directResponse`. Does that make sense?
